Database: Refactor database actions for Instances

masnax opened the pull request:
Depends on #10541.

Sorry for the size, there was a lot of collateral damage in this PR, which made for a lot of failed tests!

This is part of #10183, and refactors `db.Instance` into `cluster.Instance` and strips out fields that reference other tables: `Profiles`, `Config`, `Devices`. Using the `v2` generator, these fields are accessible via `cluster.GetInstanceConfig`, and similar functions, which are generated in the `cluster` package.

To access a "complete" Instance object, there is `InstanceToArgs` and `ToAPI`, which convert a `cluster.Instance` into a type that automatically fills `Profiles`, `Config`, and `Devices` from the database.

## Other changes
*  `InstanceToArgs`, and `instance.Instance` now hold the entire structs for the instance's list of profiles. To get the names we were querying the whole profiles table anyway, so this prevents us from having to do additional queries again every time we want to check the profile's config or devices. 
* `InstanceArgs` is now used as the instance type for `InstanceList`, as it has access to all the fields pertaining to an instance. As we are pre-fetching all instance profiles now, `InstanceList` can be simplified to just do the `InstanceToArgs` conversion, and fetch the projects. 
  * This is slightly less efficient than before, since we are calling `ToAPI` on the same profile multiple times, but I chose to go with this simpler version to make the PR more easily digestible. I think once this is through, we can add a helper that creates a list of `InstanceArgs` in a more efficient manner.
* `api.Instance` now has an `api.URL` method for getting its UsedBy url, rather than having this live in the `db` package.
* Some now-unused functions / types are removed. A few more will be removed once `Snapshots` is refactored. That's ready to go once this is merged.

tomponline commented on lxd/db/instances.go:
```diff
@@ -299,59 +299,62 @@ var ErrInstanceListStop = fmt.Errorf("search stopped")
 		}
 
-		profiles, err := cluster.GetProfiles(ctx, tx.tx, cluster.ProfileFilter{})
+		projects, err := cluster.GetProjects(ctx, tx.tx, cluster.ProjectFilter{})
```
Can we avoid this `GetProjects` call by changing the for loop below to iterate over the project names generated in `projectsByName` above? Seeing as we are just populating those entries here.

masnax replied:
Doing this would result in more database hits as we would be talking to SQL once for each iteration of `projectsByName` rather than just taking everything at once and filtering it in `go`.

tomponline replied:
Right I see, so you're saving queries by potentially transferring more data from the DB (probably the right choice).
Makes sense.

tomponline commented on lxd/db/cluster/profiles.go:
```diff
@@ -88,3 +89,52 @@ func GetProfileIfEnabled(ctx context.Context, tx *sql.Tx, projectName string, na
+	for i, profile := range profiles {
+		profileDevices[i] = config.NewDevices(profile.Devices)
+	}
+	for i := range profileDevices {
```
Extra new line needed here.

tomponline commented on lxd/backup/backup_config_utils.go:
```diff
@@ -38,11 +41,34 @@ func ConfigToInstanceDBArgs(c *config.Config, projectName string) *db.InstanceAr
+		err := state.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+			inst.Profiles = make([]api.Profile, 0, len(c.Container.Profiles))
+			for _, profileName := range c.Container.Profiles {
+				profile, err := cluster.GetProfileIfEnabled(ctx, tx.Tx(), projectName, profileName)
```
We could avoid querying the project features for every profile (via `ProjectHasProfiles` inside `GetProfileIfEnabled`) if we did it ourselves once here, and then called `GetProfile` with the right project instead.

tomponline commented on lxd/instance.go:
```diff
@@ -88,14 +89,36 @@ func instanceCreateFromImage(d *Daemon, r *http.Request, args db.InstanceArgs, h
+		if args.Profiles == nil {
+			args.Profiles = make([]api.Profile, 0, len(img.Profiles))
+			for _, name := range img.Profiles {
+				profile, err := dbCluster.GetProfileIfEnabled(ctx, tx.Tx(), args.Project, name)
```
Same comment as above about GetProfileIfEnabled, it feels like we should avoid using this function inside a loop as it will cause duplicated project feature queries.

tomponline commented on lxd/instance_patch.go:
```diff
@@ -162,7 +168,22 @@ func instancePatch(d *Daemon, r *http.Request) response.Response {
+	apiProfiles := make([]api.Profile, 0, len(req.Profiles))
 	err = d.db.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+		for _, name := range req.Profiles {
+			profile, err := cluster.GetProfileIfEnabled(ctx, tx.Tx(), projectName, name)
```
Avoid GetProfileIfEnabled call in loop.

tomponline commented on lxd/instance_put.go:
```diff
@@ -107,7 +108,22 @@ func instancePut(d *Daemon, r *http.Request) response.Response {
+		apiProfiles := make([]api.Profile, 0, len(configRaw.Profiles))
 		err = d.db.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+			for _, name := range configRaw.Profiles {
+				profile, err := cluster.GetProfileIfEnabled(ctx, tx.Tx(), projectName, name)
```
Avoid GetProfileIfEnabled in loop

tomponline commented on lxd/instances_post.go:
```diff
@@ -250,20 +266,49 @@ func createFromMigration(d *Daemon, r *http.Request, projectName string, req *ap
+		for _, name := range req.Profiles {
+			profile, ok := profilesByName[name]
+			if !ok {
+				return fmt.Errorf("Requested profile '%s' doesn't exist", name)
```
We use %q for quoting not '%s'

tomponline commented on lxd/db/cluster/instance_profiles.go:
```diff
@@ -1,3 +1,5 @@
+//go:build linux && cgo && !agent
```
Does this mean that lxd-agent is importing lxd/db/cluster package? If so this seems like a bug?

tomponline reviewed (changes requested):
Phew! That was a big one. Overall looks good to me.
However I think we should try and remove `GetProfileIfEnabled` entirely as it seems to foster inefficient query approaches.

tomponline commented on lxd/db/instances.go:
```diff
@@ -299,59 +299,62 @@ var ErrInstanceListStop = fmt.Errorf("search stopped")
-			profilesByName, ok := profilesByProjectAndName[profile.Project]
+		// Index of all projects by name and record which projects have the profiles feature.
+		for _, project := range projects {
+			// Skip any projects not associated with an instance.
```
// Skip any projects not associated with an instance in this filtered list.

masnax commented:
@tomponline I've addressed your comments: `GetProfileIfEnabled` is now `GetProfilesIfEnabled`, which absorbs the for loops that `GetProfileIfEnabled` was mostly called in and only checks the project for the profiles feature once. I didn't want to remove the helper entirely because I think it's helpful to have as a reference that `ProjectHasProfiles` should be checked when fetching profiles for a specific project.

Also added back the `&& !agent` argument to the build constraints that I just removed :)

tomponline commented:
> @tomponline I've addressed your comments: `GetProfileIfEnabled` is now `GetProfilesIfEnabled`, which absorbs the for loops that `GetProfileIfEnabled` was mostly called in and only checks the project for the profiles feature once. I didn't want to remove the helper entirely because I think it's helpful to have as a reference that `ProjectHasProfiles` should be checked when fetching profiles for a specific project.
> 
> Also added back the `&& !agent` argument to the build constraints that I just removed :)

Thanks sounds good.

tomponline commented:
jenkins: test this please

tomponline commented:
jenkins: test this please

tomponline reviewed (approved):
LGTM, I want to merge this...if I can get Jenkins to download it from github :)

tomponline commented:
I think this is as good as we're going to get it given the issues with Jenkins runners in Canonical's Boston DC, there's a known network issue ATM.
